// @discordjs/rest/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const CDN = require('./lib/CDN.cjs');
const DiscordAPIError = require('./lib/errors/DiscordAPIError.cjs');
const HTTPError = require('./lib/errors/HTTPError.cjs');
const RateLimitError = require('./lib/errors/RateLimitError.cjs');
const RequestManager = require('./lib/RequestManager.cjs');
const REST = require('./lib/REST.cjs');
const constants = require('./lib/utils/constants.cjs');
const utils = require('./lib/utils/utils.cjs');



exports.CDN = CDN.CDN;
exports.DiscordAPIError = DiscordAPIError.DiscordAPIError;
exports.HTTPError = HTTPError.HTTPError;
exports.RateLimitError = RateLimitError.RateLimitError;
exports.RequestManager = RequestManager.RequestManager;
exports.RequestMethod = RequestManager.RequestMethod;
exports.REST = REST.REST;
exports.ALLOWED_EXTENSIONS = constants.ALLOWED_EXTENSIONS;
exports.ALLOWED_SIZES = constants.ALLOWED_SIZES;
exports.ALLOWED_STICKER_EXTENSIONS = constants.ALLOWED_STICKER_EXTENSIONS;
exports.DefaultRestOptions = constants.DefaultRestOptions;
exports.DefaultUserAgent = constants.DefaultUserAgent;
exports.RESTEvents = constants.RESTEvents;
exports.makeURLSearchParams = utils.makeURLSearchParams;
exports.parseResponse = utils.parseResponse;
